waspc/src/Wasp/Analyzer/Parser/AST.hs

```diff
@@ -1,29 +1,63 @@
+{-# LANGUAGE DeriveFunctor #-}
+
 module Wasp.Analyzer.Parser.AST
   ( AST (..),
     Stmt (..),
     Expr (..),
     Identifier,
     ExtImportName (..),
+    Ctx (..),
+    withCtx,
+    ctxFromPos,
+    getCtxPos,
+    WithCtx (..),
   )
 where
 
+import Wasp.Analyzer.Parser.SourcePosition (SourcePosition)
 import Wasp.AppSpec.ExtImport (ExtImportName (..))
 
-newtype AST = AST {astStmts :: [Stmt]} deriving (Eq, Show)
+newtype AST = AST {astStmts :: [WithCtx Stmt]} deriving (Eq, Show)
 
--- Decl <declType> <name> <body>
-data Stmt = Decl Identifier Identifier Expr deriving (Eq, Show)
+data Stmt
+  = -- | Decl <declType> <declName> <declBody>
+    Decl Identifier Identifier (WithCtx Expr)
+  deriving (Eq, Show)
 
+{- ORMOLU_DISABLE -}
 data Expr
-  = Dict [(Identifier, Expr)]
-  | List [Expr]
-  | StringLiteral String
+  = Dict           [(Identifier, WithCtx Expr)]
+  | List           [WithCtx Expr]
+  | StringLiteral  String
   | IntegerLiteral Integer
-  | DoubleLiteral Double
-  | BoolLiteral Bool
-  | ExtImport ExtImportName String
-  | Var Identifier
-  | Quoter Identifier String
+  | DoubleLiteral  Double
+  | BoolLiteral    Bool
+  | ExtImport      ExtImportName String
+  | Var            Identifier
+  | Quoter         Identifier String
   deriving (Eq, Show)
+{- ORMOLU_ENABLE -}
 
 type Identifier = String
+
+data WithCtx a = WithCtx Ctx a
+  deriving (Eq, Show, Functor)
+
+withCtx :: (Ctx -> a -> b) -> (WithCtx a -> b)
+withCtx f (WithCtx ctx expr) = f ctx expr
+
+-- | Gives context to AST nodes -> e.g. source position from which they originated.
+-- TODO: Move to Analyzer.Parser (so level above)?
+-- TODO: Instead of having just SourcePosition, it would be better to have SourceRegion, since errors
+-- usually refer to a region and not just one position/char in the code.
+-- This is captured in an issue https://github.com/wasp-lang/wasp/issues/404 .
+data Ctx = Ctx
+  { ctxSourcePosition :: SourcePosition
+  }
+  deriving (Show, Eq)
+
+ctxFromPos :: SourcePosition -> Ctx
+ctxFromPos pos = Ctx {ctxSourcePosition = pos}
+
+getCtxPos :: Ctx -> SourcePosition
+getCtxPos = ctxSourcePosition
```

Comment:
I should extract all of this into `Analyzer/Parser/Ctx`. Possibly also move SourcePosition under that, not sure yet though, maybe it is better to have it standalone.

Reply:
Done